raft/src/main/java/org/apache/kafka/raft/QuorumState.java

```diff
@@ -78,8 +78,8 @@ public void initialize(OffsetAndEpoch logEndOffsetAndEpoch) throws IOException {
         } else {
             state = new FollowerState(election.epoch);
             if (election.hasLeader()) {
-                becomeFollower(election.epoch, election.leaderId());
-            } else if (election.hasVoted() && election.votedId() != localId) {
```

Comment:
This sounds like a good illegal state exception to me, maybe.

Reply:
I see the additional check as redundant. We would have become a candidate if the votedid was the same as our localId.